File: src/modules/iot_device/iot_device.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../../prisma/prisma.service';
import { CreateIoTDeviceDTO } from './dto';
import { BadRequestException } from '../../common/exception';
@Injectable()
export class IotDeviceService {
  constructor(private prisma: PrismaService) {}

  // Create new IoT device
  async createDevice(dto: CreateIoTDeviceDTO) {
    const exist = await this.prisma.ioTDevice.findUnique({
      where: { macAddress: dto.macAddress },
    });

    if (exist) {
      throw new BadRequestException('Device already exists');
    }

    const device = await this.prisma.ioTDevice.create({
      data: {
        macAddress: dto.macAddress,
        deviceName: dto.deviceName,
        lotId: dto.lotId,
        status: 'ACTIVE',
        lastSeen: new Date(),
      },
    });

    return device;
  }
}

File: internal/modules/iot_device/iot_device.service.go
package iot_device

import (
	"strings"
	"time"

	appErrors "backend/internal/common/errors"
)

type Service struct {
	repo *Repository
}

func NewService(repo *Repository) *Service {
	return &Service{repo: repo}
}

func (s *Service) CreateDevice(req CreateIoTDeviceRequest) (*IoTDevice, error) {
	req.MacAddress = strings.TrimSpace(req.MacAddress)
	req.DeviceName = strings.TrimSpace(req.DeviceName)

	exist, err := s.repo.FindByMacAddress(req.MacAddress)
	if err != nil {
		return nil, appErrors.NewInternal("Có lỗi xảy ra khi kiểm tra thiết bị")
	}

	if exist != nil {
		return nil, appErrors.NewBadRequest("Device already exists")
	}

	now := time.Now()

	device := &IoTDevice{
		MacAddress: req.MacAddress,
		DeviceName: &req.DeviceName,
		LotID:      req.LotID,
		Status:     DeviceStatusActive,
		LastSeen:   &now,
	}

	if err := s.repo.CreateDevice(device); err != nil {
		return nil, appErrors.NewInternal("Tạo thiết bị IoT thất bại")
	}

	return device, nil
}
